### PR TITLE
Fix #7, #8, #9, #10, #11 : add classification metrics

### LINKED ISSUE #7: Classification metrics: metrics.accuracy_score
Add a differentially private variant of the [metrics.accuracy_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html#sklearn.metrics.accuracy_score) method in corresponding location.

To do this you will likely need to leverage the method `mean` in the tools -> utils which is a differentially private mean function. 

Remember to keep the interfaces consistent!

### PR DESCRIPTION
**Pull request by Dhruv127**

### CODE REVIEW

**Review by jkfitzsimons — CHANGES_REQUESTED** [review]
Excellent work overall, I hope the comments are useful. With a few small changes you will have knocked off many of the medium PRs already! 💪  (we will need to write more or you can get points for proposing PRs yourself too :) ). Try to close out the open PRs by making the required changes before moving on though please so we can mark them as complete and give you the points!

**Comment by Dhruv127** [comment]
please check now , have made all the requested changes

File: diffprivlib/Classification Metrics/metrics.accuracy_score.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+from diffprivlib.accountant import BudgetAccountant
+# Define differentially private mean functions here
+
+def accuracy_score(y_true, y_pred, epsilon=1.0, bounds=None, random_state=None):
+    # Convert labels to numpy arrays if they're not already
+    y_true = np.asarray(y_true)
+    y_pred = np.asarray(y_pred)
+
+    # Calculate accuracy using the regular accuracy_score function
+    accuracy = metrics.accuracy_score(y_true, y_pred)
```
Comment (jkfitzsimons) [inline diffprivlib/Classification Metrics/metrics.accuracy_score.py:13]: This line is now redundant, we can delete.

File: diffprivlib/Classification Metrics/metrics.auc.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+from diffprivlib.accountant import BudgetAccountant
+
+
+def auc(y_true, y_score, epsilon=1.0, bins=10, range=None):
+    # Convert labels to numpy arrays if they're not already
+    y_true = np.asarray(y_true)
+    y_score = np.asarray(y_score)
+
+    # Calculate AUC using the regular auc function
+    auc = metrics.auc(y_true, y_score)
```
Comment (jkfitzsimons) [inline diffprivlib/Classification Metrics/metrics.auc.py:13]: This line is now redundant, please delete all dead code.

File: diffprivlib/Classification Metrics/metrics.average_precision_score.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+from diffprivlib.tools.utils import mean
+from diffprivlib.accountant import BudgetAccountant
+
+def average_precision_score(y_true, y_score, epsilon=1.0, bounds=None, random_state=None):
+    # Convert labels to numpy arrays if they're not already
+    y_true = np.asarray(y_true)
+    y_score = np.asarray(y_score)
+
+    # Calculate average precision using the regular function
+    avg_precision = metrics.average_precision_score(y_true, y_score)
```
Comment (jkfitzsimons) [inline diffprivlib/Classification Metrics/metrics.average_precision_score.py:12]: This is now dead code, please delete

File: diffprivlib/Classification Metrics/metrics.confusion_matrix.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    # Add differential privacy to the confusion matrix calculation
+    dp_conf_matrix = conf_matrix.copy()  # Start with a copy of the regular confusion matrix
+    
+    # Introduce differential privacy logic to perturb the confusion matrix
+    for i in range(dp_conf_matrix.shape[0]):
+        for j in range(dp_conf_matrix.shape[1]):
+            # Perturb each element of the confusion matrix using Laplace mechanism for differential privacy
+            scale = sensitivity / epsilon
+            laplace_noise = np.random.laplace(scale=scale, size=1)
+            dp_conf_matrix[i, j] += laplace_noise + dp_mean  # Incorporate dp_mean here
```
Comment (jkfitzsimons) [inline diffprivlib/Classification Metrics/metrics.confusion_matrix.py:28]: once the dp_mean is calculated it's essentially a constant. Why are you adding it to the confusion matrix here? I understand that for each element in the confusion matrix the max contribution a y_pred value could change the results by is 1 (like counting things) so you have added the laplacian noise with appropriate scale (good job), but then you add the dp_mean... what is the logic behind this? I think it's incorrect

File: diffprivlib/Classification Metrics/metrics.f1_score.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    # Calculate the differentially private mean of y_pred
+    dp_mean = mean(y_pred, bounds=bounds, epsilon=epsilon, random_state=random_state, accountant=BudgetAccountant())
+
+    # Introduce Laplace noise for differential privacy
+    sensitivity = 1.0  # Placeholder sensitivity, adjust based on your data
+    scale = sensitivity / epsilon
+    laplace_noise = np.random.laplace(scale=scale, size=1)
+
+    # Adjust the F1 score using dp_mean and Laplace noise
+    dp_f1 = f1 + dp_mean + laplace_noise  # You can adjust the combination based on your requirements
```
Comment (jkfitzsimons) [inline diffprivlib/Classification Metrics/metrics.f1_score.py:23]: This equation once again is incorrect. You are adding laplace noise directly but the sensitivity is wrong and then you are adding the dp_mean? A hint would be to get the dp confusion matrix working. Once you have that the f1 function can simply call it, get the result and use the TP, FP, FN from the result to calculate f1-score directly. Check the wikipedia for their relationship: https://en.wikipedia.org/wiki/F-score

**Review by jkfitzsimons — CHANGES_REQUESTED** [review]
There are some fundamental errors in the code which I've marked up. Not sure if you are using GPT or something on the confusion matrix and f1 scores but I'm certain it's incorrect. Hopefully the comments guide you to the correct solution :) Happy to answer questions if you get stuck, you are almost there!

**Comment by Dhruv127** [comment]
bhaiya , please check now , and don't i know if its allowed but i use the ai for some reference and to add comments , sorry for taking your so much time.
Please review it once more.

File: diffprivlib/Classification_Metrics/average_precision_score.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+def average_precision_score(y_true, y_score, epsilon=1.0, bounds=None, random_state=None):
+    y_true = np.asarray(y_true)
+    y_score = np.asarray(y_score)
+    
+    avg_precision = metrics.average_precision_score(y_true, y_score)
+
+    accountant = BudgetAccountant()
+    dp_mean = mean(y_score, bounds=bounds, epsilon=epsilon, random_state=random_state, accountant=accountant)
+
+    return avg_precision + dp_mean
```
Comment (jkfitzsimons) [inline diffprivlib/Classification_Metrics/average_precision_score.py:15]: What is going on here - you are getting the regular precision and then adding the dp_mean of the y_score..? How would this help? Seen as you have the confusion matrix already built, why not use it directly, see: https://towardsdatascience.com/accuracy-precision-recall-or-f1-331fb37c5cb9

**Review by jkfitzsimons — CHANGES_REQUESTED** [review]
I think there is one remaining bug. I hope I've been very clear how you can solve it :)

**Comment by Dhruv127** [comment]
@jkfitzsimons , is it done ?

**Comment by Dhruv127** [comment]
@jkfitzsimons  please check now , i have updated the confusion matrix . And for average_precision  , i can't find any direct method to calculate it from recall and precision , so i added Laplace noise , to the one obtained from sklearn.average_precison method . Please check it once .

File: diffprivlib/Classification_Metrics/average_precision_score.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    sensitivity=1
+
+    laplace_scale = sensitivity / epsilon
+
+    input_space = atom_domain(T=float), absolute_distance(T=float)
+    base_lap = make_base_laplace(*input_space, scale=laplace_scale)    
+
+    output = base_lap(output)
+
+    return output
```
Comment (jkfitzsimons) [inline diffprivlib/Classification_Metrics/average_precision_score.py:23]: What you are doing here is adding noise to precision as if it was count data (1,2,3,4,5) but the precision is in (0-1). You might be better off calculating the conf_matrix = metrics.confusion_matrix(y_true, y_pred), and then adding noise to TP and FP as you have (clip to 0, ie no value should be negative), and the precision = TP/(TP+FP) as in https://machinelearningmastery.com/precision-recall-and-f-measure-for-imbalanced-classification/#:~:text=Consider%20a%20model%20that%20predicts,95%20%2F%20(95%20%2B%2055)

**Review by jkfitzsimons — CHANGES_REQUESTED** [review]
Last change request hopefully :)
